test/vm-download

```diff
@@ -86,21 +86,43 @@ def download(image, force, stores):
     if not os.path.exists(image_file):
         os.symlink(os.path.abspath(dest), image_file)
 
-def prune(force):
+def prune(force, dryrun):
     now = time.time()
     targets = []
     for filename in os.listdir(IMAGES):
         path = os.path.join(IMAGES, filename)
-        if os.path.islink(path):
-            target = os.path.join(DATA, os.readlink(path))
+
+        # only consider original image entries as trustworthy sources
+        if path.endswith(".qcow2") or path.endswith(".partial"):
+            continue
+
+        # only consider links
+        if not os.path.islink(path):
+            continue
+
+        target = os.readlink(path)
+
+        # if the path isn't absolute, it can resolve to either the images directory or here (might be the same)
+        if not os.path.isabs(target):
+            targets.append(os.path.join(IMAGES, target))
+            targets.append(os.path.join(DATA, target))
+        else:
             targets.append(target)
+
     for filename in os.listdir(DATA):
         path = os.path.join(DATA, filename)
         if not force and os.lstat(path).st_mtime > now - testinfra.IMAGE_EXPIRE * 86400:
             continue
-        if os.path.isfile(path) and "." in path and path not in targets:
+        if os.path.isfile(path) and (path.endswith(".qcow2") or path.endswith(".partial")) and path not in targets:
             sys.stderr.write("Pruning {0}\n".format(filename))
-            os.unlink(path)
+            if not dryrun:
+                os.unlink(path)
+
+    # now also prune broken links
```

Comment:
I don't think we need to invoke a whole hundreds of lines of shell script here. It would be easy to check if the links are broken above, while enumerating them, and just remove broken links in a little loop here, no?